### PR TITLE
chore: add types to `crdt` package

### PR DESCRIPTION
**Pull request by TomAFrench**
This PR adds types to the `crdt` directory to make the structure of the merkle tree (really a merkle radix trie) clearer.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Can fix remaining type issues later.

**Review by j-f1 — DISMISSED** [review]
I like this, but I am also wary of merging right before a release. Let’s circle back once the first June release is out to merge.

**Comment by TomAFrench** [comment]
Cool, yeah no rush on this so feel free to merge whenever.

File: packages/loot-core/src/server/sync/index.ts
Comment (TomAFrench) [inline packages/loot-core/src/server/sync/index.ts:246]: I updated this based on the failing typecheck in the test but it seems like it's the test which is incorrect. I'll revert this change and update the test.

**Comment by TomAFrench** [comment]
Let's park this pending https://github.com/actualbudget/actual/pull/1077

**Comment by j-f1** [comment]
Something to consider cherry-picking: 5b235057

**Comment by TomAFrench** [comment]
> Something to consider cherry-picking: [5b23505](https://github.com/actualbudget/actual/commit/5b23505783fb94deae2a77a966bbc8cebdda7e3b)

Thanks, I noticed that the issue was being hidden by the `any` return types of `send/recv` so this helped get TS to spot this issue.

I'll need to do more digging in to this later as I think some of the tests have been writing invalid timestamps without us noticing so they will need fixing.

**Comment by TomAFrench** [comment]
I don't think I'll be able to meaningfully progress this over the next month or so. I can't run the tests locally on my laptop so I would need to constantly push and get CI to run the test suite for me.

**Comment by j-f1** [comment]
@TomAFrench would you be ok with me picking this up and taking it over the finish line?

**Comment by j-f1** [comment]
All of the tests pass now!

File: packages/loot-core/src/server/sync/index.ts
Comment (MatissJanis) [inline packages/loot-core/src/server/sync/index.ts:507]: **❓ question:** ‏why are we removing this? Smells to me like this is not JUST a type change, but there are some functional changes too. Which is risky.

File: packages/loot-core/src/server/sync/index.ts
Comment (MatissJanis) [inline packages/loot-core/src/server/sync/index.ts:694]: Same question

File: packages/loot-core/src/server/sync/index.ts
Comment (TomAFrench) [reply packages/loot-core/src/server/sync/index.ts:507]: We can remove this as https://github.com/actualbudget/actual/pull/1076/commits/9726e84a7fdb4009263069b91730e5abb36be79a expands the `Message` type to accept a `Timestamp` in its `timestamp` field.

The reason this was removed was that this wasn't satisfying the `Message` type which we had defined (it just wasn't erroring because `Timestamp.parse` returned `any`. We could then either expand the type or adjust the logic so when we pass this to `receiveMessages` then it's a valid `Message` and parse the timestamp later if necessary.

Comment (TomAFrench) [reply packages/loot-core/src/server/sync/index.ts:507]: I opted for the latter as we should be restricting our types where possible to avoid having to handle many different types, however we can restrict this type again in a later PR.

**Review by MatissJanis — APPROVED** [review]
LGTM on my end. I'll leave it to @j-f1 to give the final approval & merge.

Once this is merged I'll release a new version for the CRDT package.

**Review by j-f1 — APPROVED** [review]

**Comment by j-f1** [comment]
Thanks! I think this should be a major version bump since we did change the API surface a small amount.

**Pull request merged**
